### PR TITLE
Load RSF in one pass and stuff

### PR DESCRIPTION
**Pull request by chr1sdi**
part of: https://trello.com/c/q8anQW60/542-make-download-load-rsf-stream

### CODE REVIEW

**Comment by philandstuff** [comment]
i feel like my username makes me uniquely qualified to review this PR 😄

**Comment by chr1sdi** [comment]
sold! ;)

File: src/main/java/uk/gov/register/serialization/RSFExecutor.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    public RSFResult execute(RegisterSerialisationFormat rsf, Register register) {
+        // this instead of integer can be full command
+        // then when it's referenced just empty the value
+        // when checking for orphan there should not be any hash with command as those
+        // should have been emptied
+        // ?? memory consumption here ??
+
+        // for now hash and line from file?
+        // hash and rsf file line number
+        Map<String, Integer> shaRefLine = new HashMap<>();
```
Comment (michaelabenyohai) [inline src/main/java/uk/gov/register/serialization/RSFExecutor.java]: Can this be a `Map<HashValue, Integer>`?

File: src/main/java/uk/gov/register/serialization/RSFExecutor.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    private RSFResult validate(RegisterCommand command, Register register, int rsfLine, Map<String, Integer> shaRefLine) {
+        // this ugly method won't be needed when we have symlinks
+        // and won't have to rely on hashes
+
+        String commandName = command.getCommandName();
+        if (commandName.equals("add-item")) {
+            String hash = DigestUtils.sha256Hex(command.getCommandArguments().get(0).getBytes(StandardCharsets.UTF_8));
+            shaRefLine.put(HashingAlgorithm.SHA256 + ":" + hash, rsfLine);
```
Comment (michaelabenyohai) [inline src/main/java/uk/gov/register/serialization/RSFExecutor.java]: If you use `HashValue` in the dictionary then you won't need to do this concatenation - just use `new HashValue(HashingAlgorithm.SHA256, hash)`

File: src/main/java/uk/gov/register/serialization/RSFExecutor.java (hunk 1 above)
Comment (chr1sdi) [inline src/main/java/uk/gov/register/serialization/RSFExecutor.java]: didn't do that for the memory consumption concern as it's a bigger class, but can change that

File: src/main/java/uk/gov/register/service/RegisterSerialisationFormatService.java [hunk 3]
```diff
@@ -1,100 +1,63 @@
...
-            iterators = Iterators.concat(
-                    Iterators.singletonIterator(new AssertRootHashCommand(previousRegisterProof)),
-                    Iterators.transform(register.getItemIterator(totalEntries1, totalEntries2), AddItemCommand::new),
-                    Iterators.transform(register.getEntryIterator(totalEntries1, totalEntries2), AppendEntryCommand::new),
-                    Iterators.singletonIterator(new AssertRootHashCommand(nextRegisterProof)));
-        }
-
-        return new RegisterSerialisationFormat(iterators);
+    public RSFResult processRegisterSerialisationFormat(RegisterSerialisationFormat rsf) {
+        final RSFResult[] rsfResult = new RSFResult[1];
```
Comment (michaelabenyohai) [inline src/main/java/uk/gov/register/service/RegisterSerialisationFormatService.java]: Why is this an array?

**Pull request merged**